src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2013-2022 The Foundry Visionmongers Ltd
+/**
+ * Base class for all specification traits.
+ */
+#pragma once
+#include <algorithm>
+#include <memory>
+#include <utility>
+
+#include "../Specification.hpp"
+
+namespace openassetio {
+inline namespace OPENASSETIO_VERSION {
+namespace specification::trait {
+
+/**
+ * Abstract CRTP base class for specification traits.
+ *
+ * @tparam Derived Concrete subclass.
+ */
+template <class Derived>
+struct TraitBase {
+  using SpecificationPtr = std::shared_ptr<Specification>;
+
+  explicit TraitBase(SpecificationPtr spec) : spec_{std::move(spec)} {}
+
+  [[nodiscard]] static const TraitId& traitId() { return Derived::kId; }
+  [[nodiscard]] bool isValid() const { return spec_->hasTrait(traitId()); }
+
+ protected:
+  SpecificationPtr& spec() { return spec_; }
+  [[nodiscard]] const SpecificationPtr& spec() const { return spec_; }
+
+  template <class T>
+  [[nodiscard]] bool getTraitProperty(T* out, const TraitId& traitId,
+                                      const property::Key& propertyKey) const {
+    if (property::Value value; spec()->getTraitProperty(&value, traitId, propertyKey)) {
+      // TODO(DF): Make exception message more friendly.
```

Comment:
[question] Do we want to `return false` instead of excepting, if the goal of the bool/out param was to avoid exceptions in this API? Otherwise, you need both wrappers for most typical get scenarios, as missing or mistyped are potentially equally likely?

Reply:
I'm not sure about this. I didn't view the `bool` as a way to avoid all exceptions, but rather that "unset" is not an exceptional case, and is certainly not an error.

Conversely, I think a bad `T` is an exceptional error case, especially as users should be using well-behaved pre-agreed trait wrappers rather than the low-level arbitrary-dict API.  

I'm also not sure that it's a good idea to conflate "not set" with "set, but with an unexpected value type", with no way to know which.

But maybe there is a counterexample where such a case is valid and not an error?

Reply:
Jotting down an off-line discussion:

The main thing we need to figure out is that is incorrect type an "exceptional" case, or does its very possibility force consumers to always defensively code around it, mitigating some of the semantic value brought by moving `InvalidKey` to the bool-outptr pattern.

Reply:
Related to this: the current implementation will also throw if `traitId` doesn't exist in the dict (i.e. the specification doesn't support the requested trait).

Reply:
Capturing offline discussion:

The choice comes from the fact that we had previously made an API-level decision that "missing" is less important than "invalid type". Upon reflection, we can't say that for sure. Without doubt, invalid type is always bad, but in many cases missing is equally so.  Having two mechanism to categorise the failure reason is confusing for everyone, and forces an opinion on what is more or less important.

Proposal:

- `C`/`C++` Core API:  Return an `enum` instead of a `bool` to cover `ok`, `not_set` and `invalid_type`.
- Future `C++` extension: return `optional<variant<>>` to allow the host to use more standard techniques to determine the status.
- `Python`: Return `None` or value by default, add `strict` mode that throws exceptions if missing/invalid type

Reply:
Updated in b1169a1a